rfc/graph-feature-flags.md

```diff
@@ -0,0 +1,242 @@
+Instead, we compare the result of the newly introduced Subgraph with the result of the existing Monolith in terms of correctness and performance.
+
+If everything goes well, we can enable the feature flag for 1% of your users, keep monitoring,
+and gradually increase the percentage of traffic that is sent to the new Subgraph.
+If something goes wrong, we can disable the feature flag for all users without deploying a new version of the API.
+
+After rolling out the feature to 100% of your users, you can monitor it in production for a while and eventually publish the new Subgraph Schema without the feature flag.
+You can then remove the feature flag as well as the existing logic in the Monolith that is now obsolete.
+
+### Schema Evolution
+
+Feature flags allow you to publish breaking changes to your schema without breaking existing clients.
+You can rename a field and publish the new schema under a feature flag.
+In your codebase, you can maintain two versions of the Schema, one with the old field name and one with the new field name.
+You can expose the two versions of the Schema on a different path, e.g., `/v1/graphql` and `/v2/graphql`.
+Existing clients can continue to use the old Schema until they are ready to switch to the new Schema.
+
+### Dynamic Configuration
+
+Feature flags allow you to change the behavior of your API at runtime.
+Depending on the request context, e.g. through a Header, JWT Claim, or a cookie, features can be enabled or disabled.
+This allows you to create personalized experiences for different users without having to deploy multiple Routers or APIs.
+
+### Staging Environments
+
+Feature flags allow you to test new features in a staging environment before rolling them out to production,
+but more importantly, developers can test new features in a "real" environment without getting in the way of other developers.
+For larger environments, it's not possible to spin up the whole infrastructure on a developer's machine.
+If you want to test a new feature end-to-end, you need to have a complete staging environment that is as close to production as possible.
+This means that you'll very likely have one or more frontend applications, and many Subgraphs that are part of the Federated API.
+To be able to test a new feature, you need the entire stack to be running, so it's unlikely that you can run the whole stack for each feature in isolation.
+As a consequence, you need a shared staging environment where developers can test their features without affecting other developers,
+hence the need for feature flags so that individual developers can test their changes in as much isolation as possible.
+
+In a staging environment, you can enable a feature flag by setting a specific cookie or header in your client.
+
+## Summarizing the Problem
+
+If we condense the four main reasons, we can summarize the problem as two main points:
```

Comment:
It would be very valuable to draw the whole process of feature designing, roll-outs, analysis, rollbacks, and persistent. Then compare how well this can be integrated in existing CI/CD workflows.